### ISSUE TITLE
Artifact Caching proxy issue (Pipeline can't find the remote artifactory)

### ISSUE THREAD

**Issue by IBA-mainframe-dev** [jenkins-infra/helpdesk#4367]
### Service(s)

Artifactory, Artifact-caching-proxy

### Summary

Hello!
Link to the dependent comment with the beginning of the discussion: https://github.com/jenkins-infra/repository-permissions-updater/pull/4150#issuecomment-2435481135

When the automated build started on the Jenkins Infa side, the pipline fails with an error (https://ci.jenkins.io/job/Plugins/job/zdevops-plugin/job/release%252Fv1.1.0/1/console):

16:09:35 [ERROR] Failed to execute goal on project zdevops: Could not resolve dependencies for project io.jenkins.plugins:zdevops:hpi:1.1.0-38.vdfc0cff2a_304
16:09:35 [ERROR] dependency: org.zowe.sdk:zowe-kotlin-sdk:jar:0.5.0 (compile)
16:09:35 [ERROR] Could not find artifact org.zowe.sdk:zowe-kotlin-sdk:jar:0.5.0 in azure-aks-internal (http://artifact-caching-proxy.artifact-caching-proxy.svc.cluster.local:8080/)
16:09:35 [ERROR] -> [Help 1]
16:09:35 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project zdevops: Could not resolve dependencies for project io.jenkins.plugins:zdevops:hpi:1.1.0-38.vdfc0cff2a_304
16:09:35 dependency: org.zowe.sdk:zowe-kotlin-sdk:jar:0.5.0 (compile)
16:09:35 Could not find artifact org.zowe.sdk:zowe-kotlin-sdk:jar:0.5.0 in azure-aks-internal (http://artifact-caching-proxy.artifact-caching-proxy.svc.cluster.local:8080/)

However, the path to the artifact is written in pom.xml (<url>https://zowe.jfrog.io/artifactory/libs-release</url>) and locally everything runs without issues.
Could you please tell me how to fix such issue?

Thanks in advance!

Repo path to the missing artifactory from pom.xml:
```
    <repositories>
        <repository>
            <id>repo.jenkins-ci.org</id>
            <url>https://repo.jenkins-ci.org/public/</url>
        </repository>

        <repository>
            <id>zowe.jfrog.io</id>
            <url>https://zowe.jfrog.io/artifactory/libs-release</url>
        </repository>
    </repositories>
```

### Reproduction steps

1. Push any changes into https://github.com/jenkinsci/zdevops-plugin repo
2. Automated build should be triggerd on the Jenkins Infra side: https://ci.jenkins.io/job/Plugins/job/zdevops-plugin/
3. Build ends with an error due to the inability to retrieve the remote artifact (even though the path to it is specified in the pom.xml)

**Comment by dduportal** [comment]
Hello @IBA-mainframe-dev , we'l look into your request next week (we are a bit busy and understaffed this week).

Quick answer to help you: we do not allow plugins to use a repository we have not added and evaluated: that is why you see this behavior. It is expected.

Could you check if the missing dependencies are not already available in the Maven central? Or if it is a few files, you can add them in the rpeository as local file repository on short term to unblock you.

**Comment by dduportal** [comment]
Back to this issue: 

I confirm that https://zowe.jfrog.io/artifactory/libs-release is not mirrored by the Jenkins artifactory (and as such by the ACP system).

Could you give us a quick description of the dependencies required in this build please? How many dependencies, with which lifecycle?

I'm asking for documenting and to evaluate what kind of long term solution would be put in place (ACP exclusion, local artifact copy with an automatic update system or adding the zowe repository in Jenkins artifactory mirrors), considering each of these choices are a different tradeoff.

Pinging for awareness (and eventually advises) @MarkEWaite and @poddingue for developer experience, and @daniel-beck @wadeck for security related to these dependencies.

**Comment by IBA-mainframe-dev** [comment]
@dduportal hello!

Basically the only problem is with the external ```zowe.jfrog.io``` artifactory with ```zowe-kotlin-sdk```, which is available when running the build locally, but is not available when running the pipline on the Jenkins infra side.

This dependency is the core of the plugin and implements the ability to connect and execute commands on IBM z/OS mainframes, and without this dependency the plugin is nothing :) .

I see the easiest option (if I understand it correctly) is to add this zowe repository in Jenkins artifactory mirrors, so that when building in pipeline, this dependency will be loaded like the others from maven central.

And by lifecycle - this dependency will be used always, periodically increasing versioning (depends on ```zowe-kotlin-sdk``` updates)

**Comment by dduportal** [comment]
> @dduportal hello!
> 
> Basically the only problem is with the external `zowe.jfrog.io` artifactory with `zowe-kotlin-sdk`, which is available when running the build locally, but is not available when running the pipline on the Jenkins infra side.
> 
> This dependency is the core of the plugin and implements the ability to connect and execute commands on IBM z/OS mainframes, and without this dependency the plugin is nothing :) .
> 
> I see the easiest option (if I understand it correctly) is to add this zowe repository in Jenkins artifactory mirrors, so that when building in pipeline, this dependency will be loaded like the others from maven central.
> 
> And by lifecycle - this dependency will be used always, periodically increasing versioning (depends on `zowe-kotlin-sdk` updates)

It is easier on short term for us to add an exception to the `settings.xml` file used to tell Maven to use the Artifact Caching Proxy. We discussed this topic during today's infra meeting and, from our scope, it would be the way to go considering:

- The dependency, as per your explanation, is at the core of your plugin (it is not a side dependency which could be swapped by another)
- We want to unblock you
- Your plugin does not have a big enough audience justifying addition security measure (comparing to plugins such as `git` for instance)

Self documenting: The cons of adding an exception is that it will start scaling well in the future. We haven't reached at threshold yet though.

**Comment by dduportal** [comment]
Note for @MarkEWaite , as discussed during the SIG infra team meeting today, the list of exceptions is found here: https://github.com/jenkins-infra/jenkins-infra/blob/c8cee0b4a0087257cd38bcb74ae3724bce2de432/dist/profile/templates/jenkinscontroller/casc/artifact-caching-proxy.yaml.erb#L14

- It currently has 6 exceptions (less than what I said, which is a good news)

**Comment by dduportal** [comment]
Hi @IBA-mainframe-dev , I believe the problem is caused on your side.

I was going to add an exception to our system and found that we already added one for your plugin 1 year ago:

- https://github.com/jenkins-infra/helpdesk/issues/3501 => https://github.com/jenkins-infra/jenkins-infra/pull/2754

The principal branch of your plugin builds with success in ci.jenkins.io because it uses the proper `id` for the `zowe` repository:

https://github.com/jenkinsci/zdevops-plugin/blob/51b7eba6500a36970b17c8e28a5420a6208773b9/pom.xml#L333-L337

**Comment by IBA-mainframe-dev** [comment]
@dduportal Yes, you were right!
With changed repository ID, now the pipline worked correctly and the build was successful:
https://ci.jenkins.io/job/Plugins/job/zdevops-plugin/view/change-requests/job/PR-35/

Now that the changes have been merged into the main branch and the github [maven-cd / release](https://github.com/jenkinsci/zdevops-plugin/actions/runs/11582873687/job/32246835718#logs) action for release has been triggered, it's time to wait for the Jenkins hub to pick up the changes from artifactory :)

Thanks again for your help and clarification!
